File: arduino/lib/paralib/geo.cpp
#include <math.h>
#include "geo.h"

const double R = 6371000; // earth radius in meters

/**
 * Compute bearing from one point to another
 * @return bearing in radians
 */
double geo_bearing(double lat1, double lng1, double lat2, double lng2) {
  const double lat1r = to_radians(lat1);
  const double lat2r = to_radians(lat2);
  const double lngDelta = to_radians(lng2 - lng1);

  const double y = sin(lngDelta) * cos(lat2r);
  const double x = cos(lat1r) * sin(lat2r) - sin(lat1r) * cos(lat2r) * cos(lngDelta);
  return atan2(y, x);
}

/**
 * Distance between two points in meters
 */
double geo_distance(double lat1, double lng1, double lat2, double lng2) {
  const double lat1r = to_radians(lat1);
  const double lat2r = to_radians(lat2);
  const double latDelta = to_radians(lat2 - lat1);
  const double lngDelta = to_radians(lng2 - lng1);

  const double sinLat = sin(latDelta * 0.5);
  const double sinLng = sin(lngDelta * 0.5);

  const double a = (sinLat * sinLat) + (cos(lat1r) * cos(lat2r) * sinLng * sinLng);
  const double c = 2 * atan2(sqrt(a), sqrt(1 - a));

  return R * c;
}

File: simulator/ts/geo/geo.ts
import { LatLng } from "../dtypes"
import { mod360, toDegrees, toRadians } from "./trig"

const R = 6371000 // earth radius in meters

/**
 * Compute bearing from one point to another
 * @return bearing in degrees
 */
export function bearing(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const lat1r = toRadians(lat1)
  const lat2r = toRadians(lat2)
  // const latDelta = toRadians(lat2 - lat1)
  const lngDelta = toRadians(lng2 - lng1)

  const y = Math.sin(lngDelta) * Math.cos(lat2r)
  const x = Math.cos(lat1r) * Math.sin(lat2r) - Math.sin(lat1r) * Math.cos(lat2r) * Math.cos(lngDelta)
  return toDegrees(Math.atan2(y, x))
}

/**
 * Distance between two points in meters
 */
export function distance(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const lat1r = toRadians(lat1)
  const lat2r = toRadians(lat2)
  const latDelta = toRadians(lat2 - lat1)
  const lngDelta = toRadians(lng2 - lng1)

  const sinLat = Math.sin(latDelta / 2)
  const sinLng = Math.sin(lngDelta / 2)

  const a = (sinLat * sinLat) + (Math.cos(lat1r) * Math.cos(lat2r) * sinLng * sinLng)
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))

  return R * c
}
export function distancePoint(ll1: LatLng, ll2: LatLng): number {
  return distance(ll1.lat, ll1.lng, ll2.lat, ll2.lng)
}

/**
 * Moves the location along a bearing (radians) by a given distance (meters)
 */
export function moveBearing(lat_degrees: number, lng_degrees: number, bear: number, dist: number): LatLng {
  const d: number = dist / R

  const lat = toRadians(lat_degrees)
  const lng = toRadians(lng_degrees)

  // Precompute trig
  const sin_d = Math.sin(d)
  const cos_d = Math.cos(d)
  const sin_lat = Math.sin(lat)
  const cos_lat = Math.cos(lat)
  const sin_d_cos_lat = sin_d * cos_lat

  const lat2 = Math.asin(sin_lat * cos_d + sin_d_cos_lat * Math.cos(bear))
  const lng2 = lng + Math.atan2(Math.sin(bear) * sin_d_cos_lat, cos_d - sin_lat * Math.sin(lat2))

  const lat3 = toDegrees(lat2)
  const lng3 = mod360(toDegrees(lng2))

  return {lat: lat3, lng: lng3}
}
